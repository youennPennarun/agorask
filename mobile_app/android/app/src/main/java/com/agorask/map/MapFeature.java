package com.agorask.map;

import android.content.Context;

import com.facebook.react.views.view.ReactViewGroup;
import com.google.android.gms.maps.GoogleMap;

/**
 * Created by nolitsou on 10/22/16.
 */
public abstract class MapFeature extends ReactViewGroup {

    public MapFeature(Context context) {
        super(context);
    }

    public abstract Object getFeature();

    public abstract void add(GoogleMap map);

    public abstract void remove(GoogleMap map);
}
