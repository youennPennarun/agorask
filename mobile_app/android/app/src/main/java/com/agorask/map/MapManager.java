package com.agorask.map;

import android.support.annotation.Nullable;
import android.view.View;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.events.RCTEventEmitter;

import java.util.Map;

/**
 * Created by nolitsou on 10/21/16.
 */
public class MapManager extends ViewGroupManager<RNMapView> {
    public static final String REACT_CLASS = "RNMap";
    private final ReactApplicationContext reactContext;

    public MapManager(ReactApplicationContext reactContext) {
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    public RNMapView createViewInstance(ThemedReactContext context) {
        return new RNMapView(reactContext, this);
    }

    @Override
    @Nullable
    public Map getExportedCustomDirectEventTypeConstants() {
        Map<String, Map<String, String>> map = MapBuilder.of(
                "onMapReady", MapBuilder.of("registrationName", "onMapReady"),
                "onLongPress", MapBuilder.of("registrationName", "onLongPress")
        );

        return map;
    }

    @Override
    public void addView(RNMapView parent, View child, int index) {
        parent.addFeature(child, index);
    }

    @Override
    public void removeViewAt(RNMapView parent, int index) {
        parent.removeFeatureAt(index);
    }
    @Override
    public int getChildCount(RNMapView view) {
        return view.getFeatureCount();
    }
    @Override
    public View getChildAt(RNMapView view, int index) {
        return view.getFeatureAt(index);
    }

    void pushEvent(View view, String name, WritableMap data) {
        reactContext.getJSModule(RCTEventEmitter.class)
                .receiveEvent(view.getId(), name, data);
    }
}
