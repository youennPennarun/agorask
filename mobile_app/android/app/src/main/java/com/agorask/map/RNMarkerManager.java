package com.agorask.map;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

import javax.annotation.Nullable;

/**
 * Created by nolitsou on 10/22/16.
 */
public class RNMarkerManager extends ViewGroupManager<RNMarkerView> {
    public static final String REACT_CLASS = "RNMarker";

    public RNMarkerManager() {
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected RNMarkerView createViewInstance(ThemedReactContext reactContext) {
        return new RNMarkerView(reactContext);
    }

    @ReactProp(name = "coordinate")
    public void setCoordinate(RNMarkerView view, ReadableMap map) {
        view.setCoordinate(map);
    }

    @Override
    @Nullable
    public Map getExportedCustomDirectEventTypeConstants() {
        Map<String, Map<String, String>> map = MapBuilder.of(
                "onPress", MapBuilder.of("registrationName", "onPress")
        );
        return map;
    }

    @Override
    public void removeViewAt(RNMarkerView parent, int index) {
        super.removeViewAt(parent, index);
        parent.update();
    }
}
