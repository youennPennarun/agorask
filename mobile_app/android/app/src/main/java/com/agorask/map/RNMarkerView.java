package com.agorask.map;

import android.content.Context;
import android.util.Log;

import com.facebook.react.bridge.ReadableMap;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

/**
 * Created by nolitsou on 10/22/16.
 */
public class RNMarkerView extends MapFeature {

    private MarkerOptions options;
    private LatLng position;
    private Marker marker;

    public RNMarkerView(Context context) {
        super(context);
    }

    @Override
    public void add(GoogleMap map) {
        Log.d("RNMarkerView", "Add new marker");
        this.marker = map.addMarker(getOptions());
    }

    @Override
    public void remove(GoogleMap map) {
        marker.remove();
        marker = null;
    }
    public void setCoordinate(ReadableMap coordinate) {
        Log.d("RNMarkerView", "set marker coordinate");
        this.position = new LatLng(coordinate.getDouble("latitude"), coordinate.getDouble("longitude"));
        if (marker != null) {
            marker.setPosition(position);
        }
    }

    private MarkerOptions createOptions() {
        Log.d("RNMarkerView", "create marker options");
        MarkerOptions options = new MarkerOptions()
                .position(position);
        return options;
    }

    public MarkerOptions getOptions() {
        if (this.options == null) {
            this.options = createOptions();
        }
        return this.options;
    }

    public LatLng getPosition() {
        return position;
    }

    public Object getFeature() {
        return this.marker;
    }
}
