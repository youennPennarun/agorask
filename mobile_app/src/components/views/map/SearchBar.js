/* @flow */

import React, {Component, PropTypes} from 'react';
import { View, StyleSheet, Dimensions, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import {clear, updateQuery, searchVenue} from '../../../redux/actions/search';
import {MKTextField} from 'react-native-material-kit';

import Icon from 'react-native-vector-icons/MaterialIcons';

const {width} = Dimensions.get('window');

const Textfield = MKTextField.textfield()
  .withPlaceholder('Search')
  .build();

type SearchBarPropsType = {
  query: string,
  search: Function,
  clearSearch: Function,
  updateSearchQuery: Function,
  openDrawer: Function,
};

export class SearchBar extends Component {
  static propTypes = {
    query: PropTypes.string,
    search: PropTypes.func.isRequired,
    clearSearch: PropTypes.func.isRequired,
    updateSearchQuery: PropTypes.func.isRequired,
    openDrawer: PropTypes.func.isRequired,
  };

  static defaultProps = {
    quer: '',
  };
  props: SearchBarPropsType;

  _tfRef: ?MKTextField = null;
  _onChangeText(text: string) {
    this.props.updateSearchQuery(text);
  }

  _onSubmit() {
    this.props.search();
  }
  _clear() {
    this.props.clearSearch();
  }
  _renderClearButton(): ?Object {
    if (!this.props.query) return null;
    return (
      <TouchableOpacity style={styles.rightIcon}
        onPress={() => { this._clear(); }}>
        <Icon name='close'
          size={20}
          color='#212121' />
      </TouchableOpacity>
    );
  }
  render(): Object {
    return (
      <View style={styles.container} >
        <TouchableOpacity style={styles.leftIcon}
          onPress={() => { this.props.openDrawer(); }} >
          <Icon name='menu'
            size={30}
            color='#929292' />
        </TouchableOpacity>
        <Textfield style={styles.textInput}
          ref={(tf) => { this._tfRef = tf; }}
          returnKeyType='search'
          placeholder='Search'
          blurOnSubmit
          onSubmitEditing={() => { this._onSubmit(); }}
          value={this.props.query}
          onChangeText={(text: string) => { this._onChangeText(text); }} />

        {this._renderClearButton()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    width: width - 20,
    height: 45,
    backgroundColor: 'white',
    marginLeft: 10,
    marginRight: 10,
    marginTop: 5,
    elevation: 2,
    flexDirection: 'row',
    alignItems: 'center',
  },
  leftIcon: {
    marginLeft: 5,
    width: 40,
    justifyContent: 'center',
    alignItems: 'center',
  },
  textInput: {
    height: 50,
    width: width - 100,
    alignSelf: 'center',
    justifyContent: 'center',
    alignItems: 'center',
  },
  rightIcon: {
    height: 50,
    width: 40,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
  },
});

/* istanbul ignore next */
const mapStateToProps = (state: Object): Object => ({
  query: state.search.query,
});
/* istanbul ignore next */
const mapDispatchToProps = (dispatch: Function): Object => ({
  clearSearch: () => {
    dispatch(clear());
  },
  updateSearchQuery: (query) => {
    dispatch(updateQuery(query));
  },
  search: () => {
    dispatch(searchVenue());
  },
});


export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchBar);
