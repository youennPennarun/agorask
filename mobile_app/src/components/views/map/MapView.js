/* @flow */
import React from 'react';
import { View, StyleSheet } from 'react-native';

import SearchBar from './SearchBar';

import Map from '../../natives/Map';
import MapMarker from '../../natives/MapMarker';

function MapView(props: Object): Object {
  return (
    <View style={{flex: 1}}>
      <Map style={{
        ...StyleSheet.absoluteFillObject,
        backgroundColor: 'red',
      }} >
        <MapMarker coordinate={{
          latitude: 54.597285,
          longitude: -5.930120,
        }} />
      </Map>
      <SearchBar openDrawer={props.openDrawer} />
    </View>
  );
}

MapView.propTypes = {

};

export default MapView;
