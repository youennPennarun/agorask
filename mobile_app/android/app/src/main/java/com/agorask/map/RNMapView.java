package com.agorask.map;

import android.content.Context;
import android.graphics.Point;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationManager;
import android.util.Log;
import android.view.View;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RNMapView extends MapView
    implements OnMapReadyCallback, GoogleMap.OnMapLongClickListener {


    private final List<MapFeature> features = new ArrayList<>();
    private final Map<Marker, RNMarkerView> markers = new HashMap<>();
    private final MapManager mapManager;
    private final ReactApplicationContext context;
    private GoogleMap map;
    public RNMapView(ReactApplicationContext reactContext,MapManager mapManager) {
        super(reactContext);
        this.context = reactContext;

        this.mapManager = mapManager;

        super.onCreate(null);
        super.onResume();
        super.getMapAsync(this);
        Log.i("RNMapView", "getMapAsync");
    }

    @Override
    public void onMapReady(final GoogleMap map) {
        Log.i("RNMapView", "onMapReady");
        this.map = map;
        map.setOnMarkerClickListener(new GoogleMap.OnMarkerClickListener() {
            @Override
            public boolean onMarkerClick(Marker marker) {
                RNMarkerView clickedMarker = markers.get(marker);
                Log.i("RNMapView", "onMarkerClick");

                if (clickedMarker != null){
                    WritableMap event = new WritableNativeMap();
                    Log.i("RNMapView", "Push 'onPress' event");
                    mapManager.pushEvent(clickedMarker, "onPress", event);
                    return true;
                }
                return false;
            }
        });
        mapManager.pushEvent(this, "onMapReady", new WritableNativeMap());

        map.setOnMapLongClickListener(this);
        map.getUiSettings().setMyLocationButtonEnabled(true);
        map.setMyLocationEnabled(true);
        zoomOnUserLocation();

        // We need to be sure to disable location-tracking when app enters background, in-case some other module
        // has acquired a wake-lock and is controlling location-updates, otherwise, location-manager will be left
        // updating location constantly, killing the battery, even though some other location-mgmt module may
        // desire to shut-down location-services.
        LifecycleEventListener listener = new LifecycleEventListener() {
            @Override
            public void onHostResume() {
                map.setMyLocationEnabled(true);
            }
            @Override
            public void onHostPause() {
                map.setMyLocationEnabled(false);
            }

            @Override
            public void onHostDestroy() {
            }
        };

        context.addLifecycleEventListener(listener);
    }

    @Override
    public void onMapLongClick(LatLng latLng) {
        WritableMap event = new WritableNativeMap();
        event.putDouble("lat", latLng.latitude);
        event.putDouble("lng", latLng.longitude);
        Point point = map.getProjection().toScreenLocation(latLng);
        event.putDouble("x", point.x);
        event.putDouble("y", point.y);
        Log.i("RNMapView", "Push 'onLongPress' event");
        mapManager.pushEvent(this, "onLongPress", event);
    }

    public void zoomOnUserLocation() {
        if (map == null) {
            return;
        }
        LocationManager locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        Criteria criteria = new Criteria();

        Location location = locationManager.getLastKnownLocation(locationManager.getBestProvider(criteria, false));
        if (location != null) {
            map.animateCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(location.getLatitude(), location.getLongitude()), 13));

            CameraPosition cameraPosition = new CameraPosition.Builder()
                    .target(new LatLng(location.getLatitude(), location.getLongitude()))      // Sets the center of the map to location user
                    .zoom(12)                   // Sets the zoom
                    .bearing(0)                // Sets the orientation of the camera to east
                    .build();                   // Creates a CameraPosition from the builder
            map.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
        }
    }

    public void addFeature(View child, int index) {
        if (child instanceof RNMarkerView) {
            RNMarkerView marker = (RNMarkerView) child;
            marker.add(map);
            features.add(index, marker);
            markers.put((Marker)marker.getFeature(), marker);
        } else {
            Log.e("RNMap", "Cannot add a child of class " + child.getClass());
        }
    }
    public void removeFeatureAt(int index) {
        Log.i("RNMapView", "removeFeatureAt " + index);
        MapFeature feature = features.remove(index);
        if (feature instanceof RNMarkerView) {
            markers.remove(feature.getFeature());
        }
        feature.remove(map);
    }
}