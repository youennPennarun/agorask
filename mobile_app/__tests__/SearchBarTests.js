/* global jest, expect, describe, it, navigator */

import 'react-native';
import React from 'react';

import {SearchBar} from '../src/components/views/map/SearchBar';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

it('renders correctly when no query', () => {
  const wrapper = shallow(
    <SearchBar clearSearch={() => {}}
      updateSearchQuery={() => {}}
      search={() => {}}
      openDrawer={() => {}} />,
  );
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('renders correctly when is filled', () => {
  const wrapper = shallow(
    <SearchBar query='Some query'
      clearSearch={() => {}}
      updateSearchQuery={() => {}}
      search={() => {}}
      openDrawer={() => {}}  />,
  );
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('Clear query and open drawer', () => {
  const clearSearch = jest.fn();
  const openDrawer = jest.fn();
  const wrapper = shallow(
    <SearchBar query='Some query'
      search={() => {}}
      updateSearchQuery={() => {}}
      clearSearch={clearSearch}
      openDrawer={openDrawer} />,
  );
  wrapper.find('TouchableOpacity').at(1).simulate('press');
  expect(clearSearch.mock.calls.length).toBe(1);
  wrapper.setProps({query: ''});
  wrapper.find('TouchableOpacity').at(0).simulate('press');
  expect(openDrawer.mock.calls.length).toBe(1);
});
