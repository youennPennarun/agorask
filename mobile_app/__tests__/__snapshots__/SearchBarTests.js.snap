exports[`test renders correctly when is filled 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "elevation": 2,
      "flexDirection": "row",
      "height": 45,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 5,
      "width": 730,
    }
  }>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginLeft": 5,
        "width": 40,
      }
    }>
    <Icon
      allowFontScaling={false}
      color="#929292"
      name="menu"
      size={30} />
  </TouchableOpacity>
  <BuiltTextfield
    blurOnSubmit={true}
    highlightColor="#3F51B5"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search"
    returnKeyType="search"
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "height": 50,
        "justifyContent": "center",
        "width": 650,
      }
    }
    underlineEnabled={true}
    value="Some query" />
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "height": 50,
        "justifyContent": "center",
        "width": 40,
      }
    }>
    <Icon
      allowFontScaling={false}
      color="#212121"
      name="close"
      size={20} />
  </TouchableOpacity>
</View>
`;

exports[`test renders correctly when no query 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "elevation": 2,
      "flexDirection": "row",
      "height": 45,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 5,
      "width": 730,
    }
  }>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginLeft": 5,
        "width": 40,
      }
    }>
    <Icon
      allowFontScaling={false}
      color="#929292"
      name="menu"
      size={30} />
  </TouchableOpacity>
  <BuiltTextfield
    blurOnSubmit={true}
    highlightColor="#3F51B5"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search"
    returnKeyType="search"
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "height": 50,
        "justifyContent": "center",
        "width": 650,
      }
    }
    underlineEnabled={true} />
</View>
`;
